docker-images/syntax-highlighter/crates/scip-syntax/src/locals.rs

```diff
@@ -1,504 +1,721 @@
-            let identifier = match node.utf8_text(source_bytes) {
-                Ok(identifier) => identifier,
-                Err(_) => continue,
-            };
-
-            references.push(Reference {
-                range: node.into(),
-                group,
-                identifier,
-                node,
+            self.add_refs_while(current_scope, &mut references_iter, |ref_capture| {
+                ref_capture.node.start_byte() < scope.start_byte()
             });
-        } else {
-            let scope = match scope {
-                Some(scope) => scope,
-                None => continue,
-            };
 
-            scopes.push(Scope::new(scope.node));
+            let new_scope =
+                self.arena
+                    .alloc(Scope::new(scope_ty.to_string(), scope, Some(current_scope)));
+            self.get_scope_mut(current_scope).children.push(new_scope);
+            current_scope = new_scope
         }
-    }
 
-    let mut root = Scope::new(root_node);
+        // We need to climb back to the top level scope and add
+        // all remaining definitions
+        loop {
+            let scope_end_byte = self.end_byte(current_scope);
+            self.add_defs_while(current_scope, &mut definitions_iter, |def_capture| {
+                def_capture.node.start_byte() < scope_end_byte
+            });
+            self.add_refs_while(current_scope, &mut references_iter, |ref_capture| {
+                ref_capture.node.start_byte() < scope_end_byte
+            });
 
-    // Sort smallest to largest, so we can pop off the end of the list for the largest, first scope
-    scopes.sort_by_key(|m| {
-        (
-            std::cmp::Reverse(m.range.start_line),
-            m.range.end_line - m.range.start_line,
-            m.range.end_col - m.range.start_col,
-        )
-    });
+            if current_scope == top_scope {
+                break;
+            }
 
-    let capacity = lvalues.len() + references.len();
+            current_scope = self.parent(current_scope)
+        }
 
-    // Add all the scopes to our tree
-    while let Some(m) = scopes.pop() {
-        root.insert_scope(m);
+        assert!(
+            definitions_iter.next().is_none(),
+            "Should've entered all definitions into the tree"
+        );
     }
 
-    while let Some(m) = lvalues.pop() {
-        root.insert_lvalue(m);
+    /// Walks up the scope tree and tries to find the definition for a given name
+    fn find_def(
+        &self,
+        scope: ScopeRef<'a>,
+        name: Name,
+        start_byte: usize,
+    ) -> Option<&Definition<'a>> {
+        let mut current_scope = scope;
+        loop {
+            let scope = self.get_scope(current_scope);
+            if let Some(def) = scope.find_def(name, start_byte) {
+                break Some(def);
+            } else if let Some(parent_scope) = scope.parent {
+                current_scope = parent_scope
+            } else {
+                break None;
```

Comment:
```suggestion
                return Some(def);
            } else if let Some(parent_scope) = scope.parent {
                current_scope = parent_scope
            } else {
                return None;
```

As a defensive measure, I would also add a loop limit with an `eprintln!` if the limit is hit, in case there is a bug elsewhere that causes an infinite loop.